sql/core/src/test/scala/org/apache/spark/sql/DatasetPrimitiveSuite.scala

```diff
@@ -253,4 +256,27 @@ class DatasetPrimitiveSuite extends QueryTest with SharedSQLContext {
     checkDataset(Seq(PackageClass(1)).toDS(), PackageClass(1))
   }
 
+  test("SPARK-20254: Remove unnecessary data conversion for primitive array") {
```

Comment:
we usually don't test optimizer in end-to-end test, see `TypedFilterOptimizationSuite` as an example